lib/realtime/ghost.ex

```diff
@@ -53,31 +53,53 @@ defmodule Realtime.Ghost do
     block_waivers: []
   ]
 
-  @spec ghosts(%{Date.t() => [Block.t()]}, [Vehicle.t()], Util.Time.timestamp()) ::
+  @spec ghosts(%{Date.t() => [Run.t()]}, [Vehicle.t()], Util.Time.timestamp()) ::
           [t()]
-  def ghosts(blocks_by_date, vehicles, now) do
-    blocks_with_vehicles = MapSet.new(vehicles, fn vehicle -> vehicle.block_id end)
+  def ghosts(runs_by_date, vehicles, now) do
+    runs_with_vehicles = MapSet.new(vehicles, fn vehicle -> vehicle.run_id end)
 
-    blocks_by_date
-    |> Helpers.map_values(fn blocks ->
-      Enum.reject(blocks, fn block ->
-        MapSet.member?(blocks_with_vehicles, block.id)
+    runs_by_date
+    |> Helpers.map_values(fn runs ->
+      Enum.reject(runs, fn run ->
+        MapSet.member?(runs_with_vehicles, run.id)
       end)
     end)
-    |> Enum.flat_map(fn {date, blocks} ->
-      blocks
-      |> Enum.map(fn block ->
-        ghost_for_block(block, date, now)
+    |> Enum.flat_map(fn {date, runs} ->
+      runs
+      |> Enum.map(fn run ->
+        ghost_for_run(run, date, now)
       end)
       |> Enum.filter(& &1)
     end)
   end
 
-  @spec ghost_for_block(Block.t(), Date.t(), Util.Time.timestamp()) :: t() | nil
-  def ghost_for_block(block, date, now) do
+  @spec ghost_for_run(Run.t(), Date.t(), Util.Time.timestamp()) :: t() | nil
+  def ghost_for_run(run, date, now) do
     now_time_of_day = Util.Time.time_of_day_for_timestamp(now, date)
 
-    case current_trip(block.trips, now_time_of_day) do
+    current_piece_trips =
+      run
+      |> Run.pieces()
+      |> Enum.find(fn piece ->
+        piece.start_time < now_time_of_day and piece.end_time > now_time_of_day
+      end)
+      |> case do
+        %Schedule.Minischedule.Piece{start_mid_route?: %{trip: trip}, trips: trips} ->
+          [trip | trips]
+
+        %Schedule.Minischedule.Piece{trips: trips} ->
+          trips
+
+        _ ->
+          []
+      end
+      # exclude AsDirected
+      |> Enum.filter(&is_binary(&1))
```

Comment:
I'm not clear on the relationship between `is_binary` and as-directed trips.

Reply:
The type of the `trips` value on `Piece` is `[Trip.id() | Trip.t() | AsDirected.t()]`. These are non-hydrated runs so none of them will be `Trip.t()`, and the trip IDs will be binaries as opposed to structs. I guess I could do an `Enum.filter` with a `match?` which would be more verbose but hopefully more readable. Let me know what you think.

Reply:
yeah I think let's do the more verbose version, it's not _terribly_ more verbose and it will be a good deal clearer